src/lib/logical_query_plan/join_node.hpp

```diff
@@ -52,9 +52,33 @@ class JoinNode : public EnableMakeForLQPNode<JoinNode>, public AbstractLQPNode {
 
   const std::vector<std::shared_ptr<AbstractExpression>>& join_predicates() const;
 
+  /**
+   * @returns true if the JoinNode has as a semi reduction role in the LQP. If so, it was most likely created by the
+   *          SemiJoinReductionRule.
+   */
+  bool is_reducer() const;
+
+  /**
+   * @returns a shared pointer to the semi join reduction's corresponding JoinNode. If the pointer is not stored in a
+   *          member variable, the LQP is traversed upwards until the corresponding JoinNode has been found, otherwise
+   *          the function fails.
+   * @pre JoinNode is set to JoinMode::Semi, and ::mark_as_reducer_of has been called.
+   */
+  std::shared_ptr<JoinNode> get_or_find_corresponding_join_node() const;
+
+  /**
+   * Sets the `is_reducer` property of this JoinNode to true, and stores a weak pointer to the
+   * @param corresponding_join_node having the same join predicate.
+   * Note: This function is meant to be called by the SemiJoinReductionRule, which adds semi join reductions to LQPs.
+   */
+  void mark_as_reducer_of(const std::shared_ptr<JoinNode>& corresponding_join_node);
+
   JoinMode join_mode;
 
  protected:
+  bool _is_reducer = false;
+  mutable std::weak_ptr<JoinNode> _corresponding_join_node;
```

Comment:
Can we find a better that notes that states that we have a "normal" join node  and a pre-filtering semi join node which we artificially add? Maybe `_corresponding_semi_join_reducer_node`. That might make it clearer that this properties might not be used for "normal" join nodes.

Reply:
The corresponding node is the reduced join, right (not the reducer)? So `_reduced_join_node`?

Reply:
`_reduced_join_node` is a good suggestion, thanks!